src/main/java/com/flowingcode/vaadin/addons/twincolgrid/TwinColGrid.java

```diff
@@ -455,46 +454,49 @@ private void updateSelection(final Set<T> addedItems, final Set<T> removedItems)
 		});
 	}
 
-	@SuppressWarnings("unchecked")
-	private void configDragAndDrop(final Grid<T> sourceGrid, final Grid<T> targetGrid) {
+	private void configDragAndDrop(final TwinColModel<T> sourceModel, final TwinColModel<T> targetModel) {
 
 		final Set<T> draggedItems = new LinkedHashSet<>();
 
-		sourceGrid.setRowsDraggable(true);
-		sourceGrid.addDragStartListener(event -> {
-			draggedGrid = sourceGrid;
-			if (!(draggedGrid.getSelectionModel() instanceof GridNoneSelectionModel)) {
+		sourceModel.grid.setRowsDraggable(true);
+		sourceModel.grid.addDragStartListener(event -> {
+			draggedGrid = null;
+
+			if (!(sourceModel.grid.getSelectionModel() instanceof GridNoneSelectionModel)) {
 				draggedItems.addAll(event.getDraggedItems());
 			}
-			targetGrid.setDropMode(GridDropMode.ON_GRID);
+			targetModel.grid.setDropMode(GridDropMode.ON_GRID);
 		});
 
-        sourceGrid.addDragEndListener(event -> {
-          if (droppedInsideGrid) {
-            droppedInsideGrid = false;
-            if (draggedGrid == null) {
-              draggedItems.clear();
-              return;
-            }
-            final ListDataProvider<T> dragGridSourceDataProvider =
-                (ListDataProvider<T>) draggedGrid.getDataProvider();
-            dragGridSourceDataProvider.getItems().removeAll(draggedItems);
-            dragGridSourceDataProvider.refreshAll();
+        sourceModel.grid.addDragEndListener(event -> {
+          boolean sourceDiffTarget = sourceModel.equals(draggedGrid);
+
+          if (targetModel.droppedInsideGrid && sourceDiffTarget) {

```

Comment:
```suggestion
          if (targetModel.droppedInsideGrid && sourceModel.grid==draggedGrid) {

```
(compare instances, the operation is clearer than the variable name and has almost the same length, draggedGrid is now a Grid - per the other suggestion)